### ISSUE TITLE
[Bug] resolution syntax inconsistently processed

### ISSUE THREAD

**Issue by eps1lon** [yarnpkg/berry#402]
**Describe the bug**

Using `resolutions` to resolve to a completely different packages no longer works.

**To Reproduce**

The following (part of a) package.json
```json
  "resolutions": {
    "**/@babel/plugin-transform-destructuring": "npm:@minh.nguyen/plugin-transform-destructuring@^7.5.2"
  },
````

used to create the following lockfile entry in v1:
```yaml
"@babel/plugin-transform-destructuring@^7.5.0", "@babel/plugin-transform-destructuring@npm:@minh.nguyen/plugin-transform-destructuring@^7.5.2":
  version "7.5.2"
  resolved "https://registry.yarnpkg.com/@minh.nguyen/plugin-transform-destructuring/-/plugin-transform-destructuring-7.5.2.tgz#49de3e25c373fadd11471a2fc99ec0ce07d92f19"
  integrity sha512-DIzWFKl5nzSk9Hj9ZsEXAvvgHiyuAsw52queJMuKqfZOk1BOr9u1i2h0tc6tPF3rQieubP+eX4DPLTKSMpbyMg==
  dependencies:
    "@babel/helper-plugin-utils" "^7.0.0"
```
In yarn v2 a `yarn install` errors with 
```
➤ YN0000: ┌ Resolution step
➤ YN0016: │ @babel/plugin-transform-destructuring@npm:7.5.2: Registry failed to return reference "7.5.2"
➤ YN0000: └ Completed in 15.97s
➤ YN0000: Failed with errors in 15.97s
```

**Screenshots**
N/A

**Environment if relevant (please complete the following information):**

 - OS: Ubuntu 19.04
 - Node version: 10.16.0
 - Yarn version: 2.0.0-rc.1

**Additional context**

`yarn` ran with a lockfile that was generated in v1. Removing the lockfile and letting `yarn@2.x` create a completely new lockfile did not produce the same error. It will, however, ignore the resolution field.

Changing the entry to `"@babel/plugin-transform-destructuring": "npm:@minh.nguyen/plugin-transform-destructuring@^7.5.2",` restored the old behavior. Were the old entries unintentionally working or the syntax changed?

The use case is to force dependencies to use a fork of a package (here `next` should use the fork instead) without having to use webpack or babel aliases (especially hard to get working in `next`).

**Comment by eps1lon** [comment]
Update: Once I changed the resolutions entry I could revert back to the old syntax without the lockfile entry changing

**Comment by arcanis** [comment]
**tldr:** The reason for that is that when we designed the resolution field in Yarn 1 I wasn't experienced enough to notice an inherent flaw, and the v2 fixes that issues. Using glob patterns is discouraged because Yarn only cares about a single level of nesting anyway.

The flaw is that, in Yarn's model, one dependency always has the same set of dependencies (except for the special case of peer dependencies). This is different from the npm model where one dependency may have multiple effective dependency sets depending on its location in the dependency tree.

When we designed the `resolutions` field, I didn't take that into account and thus suggested to use a glob-like syntax to describe multiple levels of specialization: "change the dependency `foo` of `bar` if `bar` is a dependency of `baz`". This doesn't make sense in the Yarn model, because changing the dependencies of `bar` will change them for every instance of `bar` - not only those that are dependencies of `baz`.

So in the v2, the syntax is now more clearly defined. To quote [the documentation](https://next.yarnpkg.com/configuration/manifest#resolutions) (plus a fix that I need to backport there), the following fields are accepted:

```json
{
  "resolutions": {
    "relay-compiler": "3.0.0",
    "webpack/memory-fs": "0.4.1",
    "@babel/core/json5": "2.1.0",
    "@babel/core/@babel/generator": "7.3.4",
    "@babel/core@npm:7.0.0/@babel/generator": "7.3.4",
    "@babel/core@npm:7.0.0/@babel/generator@npm:^7.0.0": "7.3.4",
  }
}
```

**Comment by eps1lon** [comment]
So adding a warning if glob syntax is detected and a changelog entry would be appropriate?

**Comment by arcanis** [comment]
Yep! We are already catching syntax problems in the manifest, the main task is to surface them to the install (maybe in a "validation step"):

https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-core/sources/Manifest.ts#L287
